### DISCUSSION TITLE
ESP32 UART Interrupt

### DISCUSSION THREAD

**Discussion by engineertrooper** [micropython/micropython#9976] [Libraries & Drivers] [answered]
hi community 👋 

**Problem:**
The ESP32 Uart interrupt is not working for all boards, like in documentation meantioned. The command ```UART.irq()``` is not supported by the [esp32 dev board](https://micropython.org/download/esp32/) with the micropython v.1.19.1.

**Info:**
I am working on a Logger project. The Logger logs received uart stream and save the data on a sd card. The sd card is working and also the stream is collected correctly. Now I want to use an interrupt to organize the executing of diffrent requierments like bluetooth connection, transfer data live via bluetooth, make leds blinking, ... blabla makes fun to build it and to try micropython out 😃 

I have a basic understanding about the [interrupt handlers](https://docs.micropython.org/en/v1.19.1/reference/isr_rules.html?highlight=event%20handler), [uasyncio](https://docs.micropython.org/en/v1.19.1/library/uasyncio.html), concept of event handler, concept of event listener and also the [observer pattern](https://refactoring.guru/design-patterns/observer/python/example).

The WWW was not helpful because there are descriptions of ```UART.irq()``` for diffrent boards, but not my board. The Lib *uasyncio* doesn't look like the solution to my problem. Because later on I want to try out [multi threads](https://docs.micropython.org/en/latest/library/_thread.html) as well 🤔 and also interrupts has a purpose 😉 

**Code snippit:**
```py
uart = UART(2, 921600)
uart.init(baudrate=921600, bits=8, parity=None, stop=1)
uartIrq = uart.irq(trigger=UART.RX_ANY, handler=handle_interrupt)
```

*Error Message:*
```
  File "main.py", line 76, in <module>
AttributeError: 'UART' object has no attribute 'irq'
MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32
```

**Question:**

- Is there an alternative to ```UART.irq()``` if esp32 is used?
- Are there pseudo "interrupts" working good in other projects with the esp32? (something tricky maybe?)
- Or do I miss something here?

Thanks for any thought and shared information ℹ️ 

## UPDATE

It looks like the Interrupt is existing, but is already used, see [here](https://forum.micropython.org/viewtopic.php?t=11104)
.... searching for a solution still, any idea?

**Comment by straga** [comment, marked as answer]
You can use uasyncio stream
coro task to wait for any data on RX - do some... , other times you can run any other tasks.
but:
https://github.com/micropython/micropython/issues/8867

**Reply by engineertrooper** [reply]
I will test it and come back with feedback.
But first try looks promising 😃  

------

Update: works fine!

**Reply by alxkos** [reply]
didnt go well for me. surely i made some mistakes. 
but i found a simple method putting an IRQ on RX pin (also disabling where i dont need it and reanabling) did the job for me.
